# Exploring eBay Kleinanzeigen Car Sales Data

In this project, we'll be cleaning data collected from German eBay site "Kleinanzeigen" to analyze used car listings. While trends and takeaways are sure to emerge, and will be outlined in the 'conclusions' section of this project, the primary objective is to guide the reader through our data cleaning steps. This serves as documentation on proper data cleaning procedures and can be used to model other data cleaning projects at our company.

First, we'll start by importing pandas and our csv. Then, we'll run a few lines of code to examine the general structure of our data.

In [1]:
import pandas as pd
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
print(autos.info())
print("\n")
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

### Initial Observations
We can see from our "autos.info()" code that there are 50,000 data points in this data set. Each entry has 19 points of information, represented by our columns. Five columns are represented as numbers, while the rest are represented as strings of information. Unforunately, we can already see that there are some errors with our data set, as not each column contains 50,000 points of information. Therefore, some rows have missing data entries for certain columns and will have to be corrected or removed. These columns, in order, are: "vehicleType", "gearbox", "model", "fueltype", "NotRepairedDamage". Other problems in our data could affect this, so let's set it aside until later.

Next, we'll look at the first 5 rows of our data set to see an example of each column. To do this, we used the "autos.head()" code in cell 3. What stands out here is some of the columns are using german words as data, while some are in english. Since our company operates in english, we'll convert the german columns to english before the end of our data cleaning. Additionally, each column uses camelcase, instead of Python's preferred snakecase format. Since this is the easiest to fix, let's start there.

In [4]:
autos.rename(columns={'dateCrawled':'date_crawled',
                      'offerType':"offer_type",
                       'vehicleType':'vehicle_type',
                       'yearOfRegistration':'registration_year',
                       'powerPS':'power_PS',
                       'monthOfRegistration':'registration_month',
                       'fuelType':'fuel_type', 
                       'notRepairedDamage':'unrepaired_damage',
                       'dateCreated':'ad_created', 
                       'nrOfPictures':'number_of_pictures', 
                       'postalCode':'postal_code', 
                       'lastSeen':'last_seen'}, inplace=True
            )
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Above, we created a matrix to input current column names that needed to be updated. In this same matrix, we assign new names to the corresponding columns. By returning the ".head()" of our data set, we can see that the change was successful.

### Correcting Numerical Data
Upon further examination, we can see that the price and the odometer columns are classified as text, rather than numbers, due to the measurement being included in each cell. To fix this, we'll need to remove the metrics ($, km) from each cell, and update the column header to include these measurements for legibility. 

In [5]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)

autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype(int)

autos.rename(columns = {"odometer":"odometer_km"}, inplace=True)


Since we're focused on numerical data at the moment, lets continue exploring the odometer and price columns. 

We're interested in any data that doesn't look right. For each column, we'll analyze the minimum and maximum values and look for any numbers that look unrealistically high or low that we might want to remove.

In [6]:
#To see how many unique values:
print(autos["price"].unique().shape)
print("\n")
#To view min, max, median, mean, etc:
print(autos["price"].describe())
print("\n")
#15 of the least expensive entries:
print(autos["price"].value_counts().sort_index(ascending=True).head(15))
print("\n")
#15 of the most expensive entries:
print(autos["price"].value_counts().sort_index(ascending=False).head(15))

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64


From our code, we learn that there are 2,357 unique data entries in our price column. Next, we are returned a few helpful labels, such as min and max. However, we can see our first red flag here. We're looking at the prices of used cars, but the data is returned in scientific notation, suggesting several large numbers are contaminating our data set. After returning the 15 least and most expensive entries, our suspicions are confirmed. While the ~1500 zero and one dollar entries aren't surprising (ebay is an auction site, after all), we can see around 10 excessively large numbers are skewing our data. Anything over the $350,000 mark seems unnecessary, so we'll refine our data with a single line of code below:

In [7]:
autos = autos[autos["price"].between(1,350000)]
print(autos["price"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


Success! Usable data! 

Now that we've successfully removed any errors in our price column, lets take a look at our odometer column.

In [8]:
#To see how many unique values:
print(autos["odometer_km"].unique().shape)
print("\n")
#To view min, max, median, mean, etc:
print(autos["odometer_km"].describe())
print("\n")
#15 of the least expensive entries:
print(autos["odometer_km"].value_counts().sort_index(ascending=True).head(15))
print("\n")
#15 of the most expensive entries:
print(autos["odometer_km"].value_counts().sort_index(ascending=False).head(15))

(13,)


count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64


150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64


Nothing here looks out of the ordinary, so we can move on.

### Correcting Date Data
We'll first look at date_crawled, last_seen, and ad_created columns, as they are represented by strings, to find any errors in our data. These 3 columns were populated by the web-crawler that collected this data, so they are represented in date-time format. Since the time isn't necessary for our analysis, we can shorten our focus to the first 10 characters of each string (just the date.)

In [9]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The main takeaway from the date crawled column is the uniformity of these listings. there was between one and three percent of our listings populated each day of the month, with the exception being the last two days of the crawling. During these days, less than 1% of our listings were seen. The sample size is small, so an exact reason to pin down, but the strongest possibility is that the crawler had indexed the entire website by then, so the last two days were _only_ the new listings created on the site that day. Taking the average of these two days might suggest the average number of daily listings created, but that isn't a metric we're interested in at the moment.

In [10]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

On the contrary, in the "last_seen" column, over 47% of our listings were last seen in the last 3 days of the crawler indexing the site. The trend here is a slow growth to ~2%, which is probably the daily burn for car listings.

In [11]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

The "ad_created" column wasn't generated by the crawler, so any errors would not stem from that possibility. Regardless, the major trend is that more ads were created recently, which makes sense. The website exists to sell cars, so if the majority of _current_ listings were created more than 30 days in the past, the data would suggest the website is a poor way to sell cars.

Next, we'll look at the "registration_year" column to explore the data for errors:

In [12]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

First of all, we can see that not all 50,000 rows of our set are populated. That's ok for now, we can choose what to do with missing entries after we correct anomalies across our data set. But with a minimum of 1000 and a max of 9999, we can see that there are definitely errors in the column for registration year. 

Since our data was scraped in 2016, its safe to assume that any data that falls after 2016 is an error and can be removed. Likewise, any vehicles listed before 1900 would be powered by _actual_ horses.

Below, we'll use similar code we used to limit our "price" column.

In [13]:
autos = autos[autos["registration_year"].between(1900,2016)]
#To output the now corrected data:
autos["registration_year"].value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002035
          ...   
1966    0.000471
1976    0.000450
1969    0.000407
1975    0.000386
1965    0.000364
1964    0.000257
1963    0.000171
1959    0.000129
1961    0.000129
1910    0.000107
1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1951    0.000043
1934    0.000043
1957    0.000043
1955    0.000043
1953    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.000021
1948    0.000021
1927    0.0000

### Exploring by Car Brands
Next, we'll take a look at the "brands" column. First, we'll want to display the information. Then we can decide what to do based on what the data shows us:

In [14]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We'll be focusing on finding the mean sales data for each brand. Since there are many different brands of cars, we'll be focusing on the brands that our customers buy and sell the most. Therefore, we'll be focusing our analysis only on brands that have greater than 5% of sales on our website.

In [15]:
brands = autos['brand'].value_counts(normalize = True)
most_common_brands = brands[brands > 0.05].index
most_common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [16]:
brands_mean_price = {}

for car in most_common_brands:
    brand = autos[autos['brand'] == car]
    mean_price = brand['price'].mean()
    brands_mean_price[car] = int(mean_price)
    
brands_mean_price

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

Interestingly, the cars that sell the most frequently are not also the brands that have the highest mean sale point. Using the data above, the company could restructure its marketing efforts to focus on owners of one of the above brands. One last point to observe is a bimodal distribution. Of the most frequently sold cars on our website, they could be classified as either a 'budget' or 'luxury' brand. But there could be another factor - average odomoter readings. 

In [17]:
brands_mean_mileage = {}

for car in most_common_brands:
    brand = autos[autos['brand'] == car]
    mean_mileage = brand['odometer_km'].mean()
    brands_mean_mileage[car] = int(mean_mileage)

brands_mean_mileage

{'audi': 129157,
 'bmw': 132572,
 'ford': 124266,
 'mercedes_benz': 130788,
 'opel': 129310,
 'volkswagen': 128707}

In [19]:
price_series = pd.Series(brands_mean_price)
mileage_series = pd.Series(brands_mean_mileage)
top_brand_info = pd.DataFrame(price_series, columns=['mean_price'])
top_brand_info['odomoter_km'] = mileage_series
top_brand_info

mean_price  odomoter_km
audi                 9336       129157
bmw                  8332       132572
ford                 3749       124266
mercedes_benz        8628       130788
opel                 2975       129310
volkswagen           5402       128707

### Conclusion 
From the data above, we can see that the mean odometer readings are close to the same. This tells us that the variations in mean price are correlated to variation in car brands, and not due to the vehicles mileage.

There is (always) more cleaning that could be done to the data set, but what has been performed so far sets a good outline for data cleaning steps that should be followed at our company.